fs-repo-11-to-12/migration/swapper.go

```diff
@@ -177,10 +180,123 @@ func (cswap *CidSwapper) prepareWorker(resultsCh <-chan query.Result) (uint64, u
 	return sw.swapped, errored
 }
 
+func (cswap *CidSwapper) swapWorkerFlatFS(fsdsPath string, fsdsShard *flatfs.ShardIdV1, swapCh <-chan Swap, reverting bool) (uint64, uint64) {
+	var swapped, errored uint64
+
+	const flatfsExtension = ".data"
+	prefix := ktds.PrefixTransform{Prefix: blocksPrefix}
+
+	getPath := func(basePath string, key ds.Key) (string, string) {
+		child := prefix.InvertKey(key)
+		noslash := child.String()[1:]
+		dir := filepath.Join(fsdsPath, fsdsShard.Func()(noslash))
+		file := filepath.Join(dir, noslash+flatfsExtension)
+
+		return dir, file
+	}
+
+	genericSwker := &swapWorker{
+		store:      cswap.Store,
+		syncPrefix: cswap.Prefix,
+	}
+
+	// Process keys from the results channel
+	for sw := range swapCh {
+		if reverting {
+			old := sw.Old
+			sw.Old = sw.New
+			sw.New = old
+		}
+
+		if !sw.Old.Parent().Equal(sw.New.Parent()) {
+			log.Error("could not swap %s->%s. The namespaces changed. Skipping.", sw.Old, sw.New)
+			errored++
+			continue
+		}
+
+		if !sw.Old.Parent().Equal(blocksPrefix) {
+			err := genericSwker.swap(sw.Old, sw.New, reverting)
+
+			// The datastore does not have the block we are planning to
+			// migrate.
+			if err == ds.ErrNotFound {
+				log.Error("could not swap %s->%s. Could not find %s even though it was in the backup file. Skipping.", sw.Old, sw.New, sw.Old)
+				continue
+			} else if err != nil {
+				log.Error("swapping %s->%s: %s", sw.Old, sw.New, err)
+				errored++
+				continue
+			}
+
+			if cswap.SwapCh != nil {
+				cswap.SwapCh <- Swap{Old: sw.Old, New: sw.New}
+			}
+			continue
+		}
+
+		_, oldPath := getPath(fsdsPath, sw.Old)
+		newDir, newPath := getPath(fsdsPath, sw.New)
+
+		_, err := os.Stat(oldPath)
+		if err != nil {
+			log.Error("could not swap %s->%s. Could not find %s even though it was in the backup file %s. Skipping.", sw.Old, sw.New, sw.Old, err.Error())
+			continue
+		}
+
+		if err := os.Mkdir(newDir, 0755); err != nil && !os.IsExist(err) {
+			log.Error("could not swap %s->%s. Skipping.", sw.Old, sw.New, err.Error())
+			continue
+		}
+
+		if err := os.Rename(oldPath, newPath); err != nil {
+			log.Error("could not swap %s->%s. Skipping.", sw.Old, sw.New, err.Error())
+			errored++
+			continue
+		}
+		swapped++
+
+		const swapLogThreshold = 10000
+		if swapped%swapLogThreshold == 0 {
+			log.Log("Migration worker has moved %d flatfs files", swapLogThreshold)
+		}
```

Comment:
I wanted to see some progress and it looks like the standard worker also logs per-worker instead of for the system as a whole so I did this here (although flatfs + standard logging are tracked separately).

Thoughts?

Reply:
@aschmahmann can you change this to log the total number of files moved (swapped)? Now it just repeats:

```
Migration worker has moved 10000 flatfs files
Migration worker has moved 10000 flatfs files
Migration worker has moved 10000 flatfs files
Migration worker has moved 10000 flatfs files
Migration worker has moved 10000 flatfs files
Migration worker has moved 10000 flatfs files
Migration worker has moved 10000 flatfs files
```

Reply:
Added some logging changes in 7bdf4b4361dff3ae04887299f560eac6489165f8

Reply:
all good ideas